utils/multi_frameworks_dataset.py

```diff
@@ -81,23 +135,54 @@ def get_tf_data(self):
             normalization=image_preprocessing,
             **self.ds_description,
         )
-        for key, split in zip(['dataset', 'test_dataset'], ['train', 'test']):
+        splits = ["test"]
+        datasets = ["test_dataset"]
+        if self.train_val_split_spec:
+            splits.append('train')
+            datasets.append('dataset')
+        else:
+            splits += [
+                f'train[:{self.ds_description["n_samples_train"]}]',
+                f'train[{self.ds_description["n_samples_train"]}:]',
+            ]
+            datasets += ["dataset", "val_dataset"]
+        for key, split in zip(datasets, splits):
             ds = tfds.load(
                 self.tf_ds_name,
                 split=split,
                 as_supervised=True,
             )
-            if key == 'test_dataset':
+            if key != "dataset":
                 ds = ds.map(
-                    lambda x, y: (image_preprocessing(x), y,),
+                    lambda x, y: (
+                        image_preprocessing(x),
+                        y,
+                    ),
                     num_parallel_calls=tf.data.experimental.AUTOTUNE,
                 )
             data_dict[key] = ds
-        return 'object', data_dict
+        if self.train_val_split_spec:
+            data_dict["val_dataset"] = filter_ds_on_indices(
+                data_dict["dataset"],
+                self.tf_val_indices,
+            ).map(
+                lambda x, y: (
+                    image_preprocessing(x),
+                    y,
+                ),
+                num_parallel_calls=tf.data.experimental.AUTOTUNE,
+            )
+            data_dict["dataset"] = filter_ds_on_indices(
+                data_dict["dataset"],
+                self.tf_train_indices,
+            )
+        return "object", data_dict
```

Comment:
With the proposed changes.
```suggestion
        
        train_idx, val_idx = self.get_train_val_indices()
        data_dict["val_dataset"] = filter_ds_on_indices(
            data_dict["dataset"], val_idx,
        ).map(
            lambda x, y: (
                image_preprocessing(x),
                y,
            ),
            num_parallel_calls=tf.data.experimental.AUTOTUNE,
        )
        data_dict["dataset"] = filter_ds_on_indices(
            data_dict["dataset"], train_idx
        )
        return data_dict
```